File: backend/app/clients/regulations.py
import logging
from typing import Optional
from datetime import datetime, timedelta

from .base import BaseAPIClient
from ..config import get_settings
from ..models.schemas import SourceItem

logger = logging.getLogger(__name__)


class RegulationsClient(BaseAPIClient):
    def __init__(self):
        settings = get_settings()
        super().__init__(base_url=settings.regulations_base_url)
        self.api_key = settings.gov_api_key

    def _get_headers(self) -> dict:
        return {
            "X-Api-Key": self.api_key,
            "Accept": "application/json",
        }

    def _normalize_document(self, doc: dict) -> SourceItem:
        attrs = doc.get("attributes", {})
        doc_id = doc.get("id", "")

        url = f"https://www.regulations.gov/document/{doc_id}"

        return SourceItem(
            source_type="regulations_document",
            id=doc_id,
            title=attrs.get("title", "Untitled Document"),
            agency=attrs.get("agencyId"),
            date=attrs.get("postedDate"),
            url=url,
            excerpt=attrs.get("summary") or attrs.get("abstract"),
        )

    def _normalize_docket(self, docket: dict) -> SourceItem:
        attrs = docket.get("attributes", {})
        docket_id = docket.get("id", "")

        url = f"https://www.regulations.gov/docket/{docket_id}"

        return SourceItem(
            source_type="regulations_docket",
            id=docket_id,
            title=attrs.get("title", "Untitled Docket"),
            agency=attrs.get("agencyId"),
            date=attrs.get("lastModifiedDate") or attrs.get("modifyDate"),
            url=url,
            excerpt=attrs.get("abstract") or attrs.get("summary"),
        )

    async def search_documents(
        self,
        search_term: str,
        date_ge: Optional[str] = None,
        date_le: Optional[str] = None,
        sort: str = "-postedDate",
        page_size: int = 10,
        page_number: int = 1,
    ) -> tuple[list[dict], list[SourceItem]]:
        params = {
            "filter[searchTerm]": search_term,
            "sort": sort,
            "page[size]": page_size,
            "page[number]": page_number,
        }

        if date_ge:
            params["filter[postedDate][ge]"] = date_ge
        if date_le:
            params["filter[postedDate][le]"] = date_le

        url = f"{self.base_url}/documents"
        logger.info(f"Searching Regulations.gov documents: {search_term}")

        data = await self._request_with_retry(
            method="GET",
            url=url,
            headers=self._get_headers(),
            params=params,
        )

        documents = data.get("data", [])
        sources = [self._normalize_document(doc) for doc in documents]

        return documents, sources

    async def get_document(
        self, document_id: str, include_attachments: bool = False
    ) -> tuple[dict, SourceItem]:
        params = {}
        if include_attachments:
            params["include"] = "attachments"

        url = f"{self.base_url}/documents/{document_id}"
        logger.info(f"Fetching Regulations.gov document: {document_id}")

        data = await self._request_with_retry(
            method="GET",
            url=url,
            headers=self._get_headers(),
            params=params if params else None,
        )

        doc = data.get("data", {})
        source = self._normalize_document(doc)

        return doc, source

    async def search_dockets(
        self,
        search_term: str,
        sort: str = "-lastModifiedDate",
        page_size: int = 10,
        page_number: int = 1,
    ) -> tuple[list[dict], list[SourceItem]]:
        params = {
            "filter[searchTerm]": search_term,
            "sort": sort,
            "page[size]": page_size,
            "page[number]": page_number,
        }

        url = f"{self.base_url}/dockets"
        logger.info(f"Searching Regulations.gov dockets: {search_term}")

        data = await self._request_with_retry(
            method="GET",
            url=url,
            headers=self._get_headers(),
            params=params,
        )

        dockets = data.get("data", [])
        sources = [self._normalize_docket(docket) for docket in dockets]

        return dockets, sources

    async def get_docket(self, docket_id: str) -> tuple[dict, SourceItem]:
        url = f"{self.base_url}/dockets/{docket_id}"
        logger.info(f"Fetching Regulations.gov docket: {docket_id}")

        data = await self._request_with_retry(
            method="GET",
            url=url,
            headers=self._get_headers(),
        )

        docket = data.get("data", {})
        source = self._normalize_docket(docket)

        return docket, source


def get_date_range(days: int) -> tuple[str, str]:
    end_date = datetime.now()
    start_date = end_date - timedelta(days=days)
    return start_date.strftime("%Y-%m-%d"), end_date.strftime("%Y-%m-%d")

File: backend-ts/src/clients/regulations.ts
import { BaseAPIClient } from "./base.js";
import { getSettings } from "../config.js";
import type { SourceItem } from "../models/schemas.js";

export class RegulationsClient extends BaseAPIClient {
    private apiKey: string;

    constructor() {
        const settings = getSettings();
        super(settings.regulationsBaseUrl);
        this.apiKey = settings.govApiKey;
    }

    private getHeaders(): Record<string, string> {
        return {
            "X-Api-Key": this.apiKey,
            "Accept": "application/json",
        };
    }

    private normalizeDocument(doc: Record<string, unknown>): SourceItem {
        const attrs = (doc.attributes || {}) as Record<string, unknown>;
        const docId = doc.id as string || "";

        const url = `https://www.regulations.gov/document/${docId}`;

        return {
            source_type: "regulations_document",
            id: docId,
            title: (attrs.title as string) || "Untitled Document",
            agency: attrs.agencyId as string || null,
            date: attrs.postedDate as string || null,
            url,
            excerpt: (attrs.summary as string) || (attrs.abstract as string) || null,
        };
    }

    private normalizeDocket(docket: Record<string, unknown>): SourceItem {
        const attrs = (docket.attributes || {}) as Record<string, unknown>;
        const docketId = docket.id as string || "";

        const url = `https://www.regulations.gov/docket/${docketId}`;

        return {
            source_type: "regulations_docket",
            id: docketId,
            title: (attrs.title as string) || "Untitled Docket",
            agency: attrs.agencyId as string || null,
            date: (attrs.lastModifiedDate as string) || (attrs.modifyDate as string) || null,
            url,
            excerpt: (attrs.abstract as string) || (attrs.summary as string) || null,
        };
    }

    async searchDocuments(options: {
        searchTerm: string;
        dateGe?: string;
        dateLe?: string;
        sort?: string;
        pageSize?: number;
        pageNumber?: number;
    }): Promise<{ documents: Record<string, unknown>[]; sources: SourceItem[] }> {
        const {
            searchTerm,
            dateGe,
            dateLe,
            sort = "-postedDate",
            pageSize = 10,
            pageNumber = 1,
        } = options;

        const params: Record<string, string | number> = {
            "filter[searchTerm]": searchTerm,
            "sort": sort,
            "page[size]": pageSize,
            "page[number]": pageNumber,
        };

        if (dateGe) {
            params["filter[postedDate][ge]"] = dateGe;
        }
        if (dateLe) {
            params["filter[postedDate][le]"] = dateLe;
        }

        const url = `${this.baseUrl}/documents`;
        console.log(`Searching Regulations.gov documents: ${searchTerm}`);

        const data = await this.requestWithRetry<{ data: Record<string, unknown>[] }>({
            url,
            headers: this.getHeaders(),
            params,
        });

        const documents = data.data || [];
        const sources = documents.map((doc) => this.normalizeDocument(doc));

        return { documents, sources };
    }

    async getDocument(options: {
        documentId: string;
        includeAttachments?: boolean;
    }): Promise<{ document: Record<string, unknown>; source: SourceItem }> {
        const { documentId, includeAttachments = false } = options;

        const params: Record<string, string> = {};
        if (includeAttachments) {
            params.include = "attachments";
        }

        const url = `${this.baseUrl}/documents/${documentId}`;
        console.log(`Fetching Regulations.gov document: ${documentId}`);

        const data = await this.requestWithRetry<{ data: Record<string, unknown> }>({
            url,
            headers: this.getHeaders(),
            params: Object.keys(params).length > 0 ? params : undefined,
        });

        const doc = data.data || {};
        const source = this.normalizeDocument(doc);

        return { document: doc, source };
    }

    async searchDockets(options: {
        searchTerm: string;
        sort?: string;
        pageSize?: number;
        pageNumber?: number;
    }): Promise<{ dockets: Record<string, unknown>[]; sources: SourceItem[] }> {
        const {
            searchTerm,
            sort = "-lastModifiedDate",
            pageSize = 10,
            pageNumber = 1,
        } = options;

        const params: Record<string, string | number> = {
            "filter[searchTerm]": searchTerm,
            "sort": sort,
            "page[size]": pageSize,
            "page[number]": pageNumber,
        };

        const url = `${this.baseUrl}/dockets`;
        console.log(`Searching Regulations.gov dockets: ${searchTerm}`);

        const data = await this.requestWithRetry<{ data: Record<string, unknown>[] }>({
            url,
            headers: this.getHeaders(),
            params,
        });

        const dockets = data.data || [];
        const sources = dockets.map((docket) => this.normalizeDocket(docket));

        return { dockets, sources };
    }

    async getDocket(docketId: string): Promise<{ docket: Record<string, unknown>; source: SourceItem }> {
        const url = `${this.baseUrl}/dockets/${docketId}`;
        console.log(`Fetching Regulations.gov docket: ${docketId}`);

        const data = await this.requestWithRetry<{ data: Record<string, unknown> }>({
            url,
            headers: this.getHeaders(),
        });

        const docket = data.data || {};
        const source = this.normalizeDocket(docket);

        return { docket, source };
    }
}

export function getDateRange(days: number): { startDate: string; endDate: string } {
    const endDate = new Date();
    const startDate = new Date();
    startDate.setDate(startDate.getDate() - days);

    return {
        startDate: startDate.toISOString().split("T")[0],
        endDate: endDate.toISOString().split("T")[0],
    };
}
